File: LeetCode/0525. Contiguous Array/solution.cpp
/**
 * 564 / 564 test cases passed.
 * Runtime: 156 ms
 * Memory Usage: 81.5 MB 
 */
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int, int> mp { {0, -1} };
        int ans = 0, sum = 0;
        for (int i = 0; i < nums.size(); ++ i) {
            nums[i] ? ++sum : --sum;
            if (mp.count(sum)) ans = max(ans, i - mp[sum]);
            else mp[sum] = i;
        }
        return ans;
    }
};

File: LeetCode/0525. Contiguous Array/solution.py
"""
564 / 564 test cases passed.
Runtime: 236 ms
Memory Usage: 19.6 MB
"""
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        mp = { 0:-1 }
        ans, s = 0, 0
        for i, num in enumerate(nums):
            s += 1 if num == 1 else -1
            if s in mp:
                ans = max(ans, i - mp[s])
            else:
                mp[s] = i
        return ans
                
